man1/kobo.1
.\" Copyright (C) 2017 Kevin Boone 
.\" Permission is granted to any individual or institution to use, copy, or
.\" redistribute this software so long as all of the original files are
.\" included, that it is not sold for profit, and that this copyright notice
.\" is retained.
.\"
.TH kobo 1 "March 2017"
.SH NAME
kobo \- a command-line utility for Kobo e-readers 
.SH SYNOPSIS
.B kobo
.RB [options]
.PP

.SH DESCRIPTION
\fIkobo\fR is simple Linux command-line utility for managing books, particularly
in EPUB and MOBI format, on Kobo e-readers. It is intended to be slightly
more convenient than simply copying and deleting files using Linux
shell commands. 

.SH EXAMPLE

.BI kobo\ -i\ my_book.epub

Install "my_book.epub" in the e-reader's storage, using the author name
as a subdirectory. The storage will be mounted if it is not already mounted,
and unmounted afterwards.

.SH "OPTIONS"
.TP
.BI -a,\-\-author
Specify a author, in cases where this can't be found in the book
meta-data. The author name is used only as the name of a subdirectory
in the e-reader's storage, and has no other significance. For organizational
purposes, it may be helpful to keep books by a particular author in
the same subdirectory. The choice of name won't affect the e-reader's
operation in any way
.LP
.TP
.BI -d,\-\-dir\ {directory} 
Specify the directory where the Kobo device is mounted, if this cannot
be determined automatically. By default, the utility will attempt to
mount the Kobo as a USB device in its default location. Specifying
a directory will override this behaviour
.LP
.TP
.BI -i,\-\-install\ {file}
Copy a file onto the Kobo device
.LP
.TP
.BI -i,\-\-list
List books on the Kobo device. Each item is shown with an number, which 
can be used as an argument to \-\-remove to delete it
.LP
.TP
.BI \-\-loglevel\ N
Sets the log verbosity, from 0 (the default) to 3. At the lowest level,
only fatal errors are displayed
.LP
.TP
.BI -n, \-\-nometa
When listing the contents of the Kobo's storage, show only filenames; do
not attempt to read meta-data from the files (which can be slow)
.LP
.TP
.BI -r, \-\-remove\ N
Remove the book with number N. Numbers start at zero. This operation cannot
be reversed 
.LP
.TP
.BI -v,\-\-version
Display version and copyright infomation
.LP

.SH NOTES

.SS Mounting the Kobo 

The Kobo devices is assumed to be connected using a USB cable, and will be
mounted as a mass storage device. On modern Linux systems, removable
devices are auto-mounted in standard locations; 
\fIkobo\fR
will use the 
\fIudisksctl\fR
utility to mount the device if it is not mounted, and will unmount it
afterwards so it can be safely removed. If your system does not support
this kind of auto-mounting, the --dir switch can be used to specify
the mount point.

.SS Cablibre compatibility

\fIkobo\fR
will try to maintain an author/title directory structure on the Kobo, as
Calibre uses. However, Calibre does not scan the actual device storage
for books, but reads a meta-data database maintained by the device. This
database is updated when the device is unmounted from the computer. 
Consequently, 
\fIkobo\fR
should work alongside Calibre, but these two utilities should not be used
at the same time, without giving the Kobo an opportunity to update 
its database.

.SS File formats

\fIkobo\fR will copy any e-book file that the Kobo supports onto the
device. However, it can only process meta-data in EPUB and MOBI files.
On other cases, it may be necessary to use the --author switch to
tell the utility which author directory to copy the file into.


.SH BUGS AND LIMITATIONS

It is probably possible for badly-formatted MOBO files crash the
utility. 


.SH AUTHOR AND LEGAL
\fIkobo\fR
is maintained by Kevin Boone, and is open source under the
terms of the GNU Public Licence, version 3.0. It contains contributions
from a number of other authors, whose details may be found in 
the source code.
There is no warranty
of any kind.


.SH FURTHER INFORMATION 
For more details, see 
http://kevinboone.net/README_kobo.html.


.\" end of file


src/kobo.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>
#include <string.h>
#include <ebookinfo/ebookinfo.h>
#include "kmslogging.h"  

/*==========================================================================
 Find the device (e.g., /dev/sdb) which represents the Kobo. If no Kobo
 device found, return NULL
==========================================================================*/
char *find_kobo_dev (void)
  {
  char *ret = NULL;

  FILE *f = popen ("udisksctl status", "r");
  if (f)
    {
    char s[256];
    BOOL done = FALSE;
    while (!done)
      {
      if (fgets (s, sizeof (s), f))
        {
        s[sizeof(s) - 1] = 0;
        if (s[strlen(s) - 1] == 10)
          s[strlen(s) -1] = 0;

        if (strcasestr (s, "kobo"))
          {
          char *t = strtok (s, " \t");
          while (t)
            {
            if (strlen (t) >= 3)
              {
              if (strncmp (t, "sd", 2) == 0)
                {
                ret = malloc (strlen (t) + 10);
                strcpy (ret, "/dev/");
                strcat (ret, t);
                done = TRUE;
                }
              }
            t = strtok (NULL, " \t");
            }
          }
        //printf ("%s\n", s);
        }
      else
        done = TRUE;
      }
    pclose (f);
    }
  return ret; 
  }


/*==========================================================================
 Find the directory on which the Kobo is mounted, if any. If no Kobo is
 mounted,  return NULL
==========================================================================*/
char *find_kobo_mount (void)
  {
  char *ret = NULL;
  char *kdev = find_kobo_dev ();
  
  if (kdev)
    {
    char cmd[256];
    snprintf (cmd, sizeof (cmd), "udisksctl info -b %s", kdev);
    FILE *f = popen (cmd, "r");
    if (f)
      {
      char s[256];
      BOOL done = FALSE;
      while (!done)
        {
        if (fgets (s, sizeof (s), f))
          {
          s[sizeof(s) - 1] = 0;
          if (s[strlen(s) - 1] == 10)
            s[strlen(s) -1] = 0;

          if (strcasestr (s, "mountpoints"))
            {
            char *p = strchr (s, '/');
            if (p)
              {
              ret = strdup (p);
              //printf ("%s\n", s);
              done = TRUE;
              }
            }
          }
        else
          done = TRUE;
        }
      pclose (f);
      }
    }

  return ret;
  }


/*==========================================================================
  returns TRUE if the file is a book that Kobo can handle
==========================================================================*/
BOOL kobo_is_book (const char *path)
  {
  BOOL ret = FALSE;
  const char *p = strrchr (path, '.');
  if (p)
    {
    const char *ext = p+1; 
    if (strcasecmp (ext, "epub") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "epub3") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "mobi") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "pdf") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "txt") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "html") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "rtf") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "cbr") == 0)
      ret = TRUE; 
    if (strcasecmp (ext, "cbz") == 0)
      ret = TRUE; 
    }
  return ret;
  }


/*==========================================================================
  kobo_mount
==========================================================================*/
void kobo_mount (const char *kdev)
  {
  char cmd[256];
  snprintf (cmd, sizeof (cmd), "udisksctl mount -b %s > /dev/null", kdev);
  system (cmd);
  }


/*==========================================================================
  kobo_unmount
==========================================================================*/
void kobo_unmount (const char *kdev)
  {
  char cmd[256];
  snprintf (cmd, sizeof (cmd), "udisksctl unmount -b %s > /dev/null", kdev);
  system (cmd);
  }




